io.c

```diff
@@ -1911,18 +1923,25 @@ rb_io_tell(VALUE io)
     return OFFT2NUM(pos);
 }
 
-static VALUE
+static off_t
 rb_io_seek(VALUE io, VALUE offset, int whence)
 {
     rb_io_t *fptr;
     off_t pos;
 
     pos = NUM2OFFT(offset);
     GetOpenFile(io, fptr);
+    return io_setpos(fptr, pos, whence);
+}
+
+static off_t
+io_setpos(rb_io_t *fptr, off_t pos, int whence)
+{
     pos = io_seek(fptr, pos, whence);
     if (pos < 0 && errno) rb_sys_fail_path(fptr->pathv);
+    if (fptr->mode & FMODE_SETENC_BY_BOM) io_set_encoding_by_bom(fptr);
```

Comment:
We should not detect BOM when we update position.

```ruby
require "tempfile"

Tempfile.create("x", encoding: "utf-8:utf-8") do |f|
  f.write("\uFEFF" "abc" "\uFEFF" "efg")
  f.close
  File.open(f.path, 'r:bom|utf-8:utf-8') do |f|
    f.pos = 6
    p f.read.unpack("U*")
  end
end
```

should output `[65279, 101, 102, 103]` but this outputs `[101, 102, 103]`.

How about this?

```diff
diff --git a/io.c b/io.c
index 3b4e25dfa1..ba19fff799 100644
--- a/io.c
+++ b/io.c
@@ -1939,7 +1939,6 @@ io_setpos(rb_io_t *fptr, off_t pos, int whence)
 {
     pos = io_seek(fptr, pos, whence);
     if (pos < 0 && errno) rb_sys_fail_path(fptr->pathv);
-    if (fptr->mode & FMODE_SETENC_BY_BOM) io_set_encoding_by_bom(fptr);
 
     return pos;
 }
@@ -2047,6 +2046,7 @@ rb_io_rewind(VALUE io)
 
     GetOpenFile(io, fptr);
     io_setpos(fptr, 0L, SEEK_SET);
+    if (fptr->mode & FMODE_SETENC_BY_BOM) io_set_encoding_by_bom(fptr);
     if (io == ARGF.current_file) {
 	ARGF.lineno -= fptr->lineno;
     }
```